File: code/pyspark/01-primer/04-azure-sql-dw/2-read-write-batch-primer.py
# Databricks notebook source
# MAGIC %md
# MAGIC # Azure SQL Datawarehouse - batch read/write primer
# MAGIC In this exercise, we will:<br>
# MAGIC 1.  **WRITE primer**: Read curated Chicago crimes data in a Delta table, from the lab on Azure Storage, and insert into Azure SQL Datawarehouse<br>
# MAGIC 2.  **READ primer**:  Read from Azure SQL Datawarehouse<br>
# MAGIC <br>
# MAGIC **Dependency:** <br>
# MAGIC Successful completion of the primer lab on Azure Blob Storage<br>
# MAGIC 
# MAGIC **Docs:**<br>
# MAGIC https://docs.databricks.com/spark/latest/data-sources/azure/sql-data-warehouse.html#

# COMMAND ----------

# MAGIC %md
# MAGIC ### 1. Review of source data

# COMMAND ----------

# MAGIC %sql
# MAGIC select * from CRIMES_DB.CHICAGO_CRIMES_CURATED;

# COMMAND ----------

# MAGIC %sql
# MAGIC select count(*) from CRIMES_DB.CHICAGO_CRIMES_CURATED;

# COMMAND ----------

# MAGIC %md
# MAGIC ### 2. Connectivity parameters

# COMMAND ----------

#JDBC URL
jdbcURL = dbutils.secrets.get(scope = "gws-sql-dw", key = "conexion-string")

#Storage account credentials for tempDir access
spark.conf.set(
  "fs.azure.account.key.gwsblobsa.blob.core.windows.net",
  dbutils.secrets.get(scope = "gws-blob-storage", key = "storage-acct-key"))

# COMMAND ----------

# MAGIC %md
# MAGIC ### 3. Azure SQL Datawarehouse - setup

# COMMAND ----------

# MAGIC %md
# MAGIC 3.1. Create a SQL DW master key <br>
# MAGIC This is a one-time activity the first time you use the SQL Datawarehouse.<br>
# MAGIC 
# MAGIC ```
# MAGIC CREATE MASTER KEY;
# MAGIC ```

# COMMAND ----------

# MAGIC %md
# MAGIC 3.2. Create table in the SQL Datawarehouse from the portal - with query explorer UI<br>
# MAGIC 
# MAGIC ```
# MAGIC DROP TABLE IF EXISTS chicago_crimes_curated_summary;
# MAGIC CREATE TABLE chicago_crimes_curated_summary
# MAGIC (case_id varchar(200),
# MAGIC primary_type varchar(200),
# MAGIC arrest_made varchar(200),
# MAGIC case_year varchar(200),
# MAGIC case_month varchar(200),
# MAGIC case_day_of_month varchar(200));
# MAGIC ```

# COMMAND ----------

# MAGIC %md
# MAGIC ### 4. Write to Azure SQL Datawarehouse

# COMMAND ----------

# MAGIC %md
# MAGIC 4.1.  Create dataframe

# COMMAND ----------

df = spark.sql("select case_id,primary_type,cast(arrest_made as string),case_year,case_month,case_day_of_month from crimes_db.chicago_crimes_curated")
df.show()

# COMMAND ----------

# MAGIC %md
# MAGIC 4.2.  Append mode to a pre-created table

# COMMAND ----------

df.write \
  .format("com.databricks.spark.sqldw") \
  .mode("append") \
  .option("url", jdbcURL) \
  .option("forwardSparkAzureStorageCredentials", "true") \
  .option("dbTable", "chicago_crimes_curated_summary") \
  .option("tempDir", "wasbs://scratch@gwsblobsa.blob.core.windows.net/sqldwbatch-tempdir") \
  .save()

# COMMAND ----------

#Repartition for control over parallelism
#df.repartition(20000).write...

# COMMAND ----------

# MAGIC %md
# MAGIC 4.3.  Create mode to a non-existent table

# COMMAND ----------

df.write \
  .format("com.databricks.spark.sqldw") \
  .option("url", jdbcURL) \
  .option("forwardSparkAzureStorageCredentials", "true") \
  .option("dbTable", "chicago_crimes_curated_summary_new") \
  .option("tempDir", "wasbs://scratch@gwsblobsa.blob.core.windows.net/sqldwbatch-tempdir") \
  .save()

# COMMAND ----------

# MAGIC %md
# MAGIC 4.4.  Overwrite mode to a pre-existing table

# COMMAND ----------

df.write \
  .format("com.databricks.spark.sqldw") \
  .mode("overwrite") \
  .option("url", jdbcURL) \
  .option("forwardSparkAzureStorageCredentials", "true") \
  .option("dbTable", "chicago_crimes_curated_summary") \
  .option("tempDir", "wasbs://scratch@gwsblobsa.blob.core.windows.net/sqldwbatch-tempdir") \
  .save()

# COMMAND ----------

# MAGIC %md
# MAGIC ### 5. Read from Azure SQL Datawarehouse

# COMMAND ----------

df = spark.read \
  .format("com.databricks.spark.sqldw") \
  .option("url", jdbcURL) \
  .option("tempDir", "wasbs://scratch@gwsblobsa.blob.core.windows.net/sqldwbatch-tempdir") \
  .option("forwardSparkAzureStorageCredentials", "true") \
  .option("query", "select primary_type as crime_type, count(*) as crime_count from chicago_crimes_curated_summary group by primary_type") \
  .load()

display(df)

# COMMAND ----------

# MAGIC %md
# MAGIC ### 6. Predicate and projection pushdown

# COMMAND ----------

# MAGIC %md https://docs.databricks.com/spark/latest/data-sources/azure/sql-data-warehouse.html#query-pushdown-into-sql-dw

# COMMAND ----------

# MAGIC %md
# MAGIC 
# MAGIC ### 7. Housekeeping

# COMMAND ----------

# MAGIC %md
# MAGIC 1) Temporary directory cleaning - Azure Blob Storage:<br>
# MAGIC You need to periodically clear the scratch directory where the connector stores data.<br>
# MAGIC https://docs.databricks.com/spark/latest/data-sources/azure/sql-data-warehouse.html#temporary-data-management

# COMMAND ----------

# MAGIC %md
# MAGIC 2) Temporary object cleaning - Azure SQL Datawarehouse:<br>
# MAGIC Due to driver crashes or accidental cluster termination, temporary objects that need clearing may not get cleared. <br>
# MAGIC The link below details housekeeping to be done.<br>
# MAGIC https://docs.databricks.com/spark/latest/data-sources/azure/sql-data-warehouse.html#temporary-object-management

File: code/scala/01-primer/03-azure-sql-db/2-read-write-batch-primer.scala
// Databricks notebook source
// MAGIC %md
// MAGIC # Azure SQL database - batch read/write primer
// MAGIC In this exercise, we will:<br>
// MAGIC 1.  **WRITE primer**: Read curated Chicago crimes data in a Delta table, from the lab on Azure Storage, and insert into Azure SQL Database<br>
// MAGIC 2.  **READ primer**:  Read from Azure SQL Database<br>
// MAGIC <br>
// MAGIC **Dependency:** <br>
// MAGIC Successful completion of the primer lab on Azure Blob Storage<br>
// MAGIC 
// MAGIC **Docs:**<br>
// MAGIC https://docs.azuredatabricks.net/spark/latest/data-sources/sql-databases.html

// COMMAND ----------

// MAGIC %md
// MAGIC ### 1. Review of source data

// COMMAND ----------

// MAGIC %sql
// MAGIC select * from CRIMES_DB.CHICAGO_CRIMES_CURATED;

// COMMAND ----------

// MAGIC %sql
// MAGIC select count(*) from CRIMES_DB.CHICAGO_CRIMES_CURATED;

// COMMAND ----------

// MAGIC %md
// MAGIC ### 2. Credentials

// COMMAND ----------

val jdbcUsername = dbutils.secrets.get(scope = "gws-sql-db", key = "username")
val jdbcPassword = dbutils.secrets.get(scope = "gws-sql-db", key = "password")

// COMMAND ----------

// MAGIC %md
// MAGIC ### 3. Driver class

// COMMAND ----------

//Class.forName("com.microsoft.sqlserver.jdbc.SQLServerDriver")
val driverClass = "com.microsoft.sqlserver.jdbc.SQLServerDriver"

// COMMAND ----------

// MAGIC %md
// MAGIC ### 4. Connectivity parameters

// COMMAND ----------

val jdbcHostname = "gws-server.database.windows.net"
val jdbcPort = 1433
val jdbcDatabase = "gws_sql_db"

// Create the JDBC URL without passing in the user and password parameters.
val jdbcUrl = s"jdbc:sqlserver://${jdbcHostname}:${jdbcPort};database=${jdbcDatabase}"

// Create a Properties() object to hold the parameters.
import java.util.Properties
val connectionProperties = new Properties()

connectionProperties.put("user", s"${jdbcUsername}")
connectionProperties.put("password", s"${jdbcPassword}")
connectionProperties.setProperty("Driver", driverClass)

// COMMAND ----------

// MAGIC %md
// MAGIC ### 5. Persist to Azure SQL Database

// COMMAND ----------

// MAGIC %md
// MAGIC ### (1) Create table and then write to table...<br>
// MAGIC Looks great!!<br>
// MAGIC Problem: The datatypes wont exactly be right, not enough control<br>
// MAGIC Better: Create table ahead of time and look at (2)

// COMMAND ----------

spark.table("crimes_db.chicago_crimes_curated")
     .write
     .jdbc(jdbcUrl, "chicago_crimes_curated", connectionProperties)

// COMMAND ----------

// MAGIC %md 
// MAGIC ### (2) Append to table
// MAGIC Awesome..but...I want to append<br>

// COMMAND ----------

import org.apache.spark.sql.SaveMode

spark.table("crimes_db.chicago_crimes_curated")
     .write
     .mode(SaveMode.Append) // <--- Append to the existing table
     .jdbc(jdbcUrl, "chicago_crimes_curated", connectionProperties)

// COMMAND ----------

// MAGIC %md ### (3) Overwrite table<br>
// MAGIC VERY COOL..but...I want to overwrite!!<br>

// COMMAND ----------

import org.apache.spark.sql.SaveMode

spark.table("crimes_db.chicago_crimes_curated")
     .write
     .mode(SaveMode.Overwrite) // <--- Overwrite the existing table
     .jdbc(jdbcUrl, "chicago_crimes_curated", connectionProperties)

// COMMAND ----------

// MAGIC %md ### (4) Parallelize writes<br>
// MAGIC SOLD..but...I want to parallelize!!<br>

// COMMAND ----------

//Repartition & persist in append/overwrite mode per your requirement
val df = spark.sql("select * from CRIMES_DB.CHICAGO_CRIMES_CURATED")
df.repartition(200000).write.mode(SaveMode.Append).jdbc(jdbcUrl, "chicago_crimes_curated", connectionProperties)

// COMMAND ----------

// MAGIC %md ### (5) Predicate pushdown

// COMMAND ----------

// Note: The parentheses are required.
val pushdown_query = "(select * from chicago_crimes_curated where primary_type ='KIDNAPPING') kidnappings"
val df = spark.read.jdbc(url=jdbcUrl, table=pushdown_query, properties=connectionProperties)
display(df)

// COMMAND ----------

// MAGIC %md ### (6) Explain plan

// COMMAND ----------

// Explain plan with no column selection returns all columns
spark.read.jdbc(jdbcUrl, "chicago_crimes_curated", connectionProperties).explain(true)

// COMMAND ----------

// Explain plan with column selection will prune columns and just return the ones specified
// Notice that only the 3 specified columns are in the explain plan
spark.read.jdbc(jdbcUrl, "chicago_crimes_curated", connectionProperties).select("case_id", "case_year", "primary_type").explain(true)

// COMMAND ----------

// You can push query predicates down too
// Notice the Filter at the top of the Physical Plan
spark.read.jdbc(jdbcUrl, "chicago_crimes_curated", connectionProperties).select("case_id", "case_year", "primary_type").where("primary_type = 'KIDNAPPING'").explain(true)

// COMMAND ----------

// MAGIC %md ### (7) Parallelize reads
// MAGIC These options specify the parallelism of the table read. lowerBound and upperBound decide the partition stride, but do not filter the rows in table. Therefore, Spark partitions and returns all rows in the table.
// MAGIC <br>We split the table read across executors on the emp_no column using the partitionColumn, lowerBound, upperBound, and numPartitions parameters.

// COMMAND ----------

val df = (spark.read.jdbc(url=jdbcUrl,
    table="chicago_crimes_curated",
    columnName="case_id",
    lowerBound=1L,
    upperBound=1000,
    numPartitions=100,
    connectionProperties=connectionProperties))
display(df)

// COMMAND ----------

// MAGIC %md ### (8) Spark SQL

// COMMAND ----------

// MAGIC %md https://docs.azuredatabricks.net/spark/latest/data-sources/sql-databases.html